llvm/lib/Support/TextEncoding.cpp

```diff
@@ -0,0 +1,353 @@
+      if (errno == E2BIG && Capacity < Result.max_size()) {
+        HandleOverflow(Capacity, Output, OutputLength, Result);
+        // Reset converter
+        iconv(ConvDesc, nullptr, nullptr, nullptr, nullptr);
+        return std::error_code();
+      } else {
+        // Some other error occured.
+        Result.resize(Output - Result.data());
+        return std::error_code(errno, std::generic_category());
+      }
+    } else {
+      // A positive return value indicates that some characters were converted
+      // in a nonreversible way, that is, replaced with a SUB symbol. Returning
+      // an error in this case makes sure that both conversion routines behave
+      // in the same way.
+      return std::make_error_code(std::errc::illegal_byte_sequence);
+    }
+  };
+
+  do {
+    // Setup the input. Use nullptr to reset iconv state if input length is
+    // zero.
+    size_t InputLength = Source.size();
+    char *Input = InputLength ? const_cast<char *>(Source.data()) : nullptr;
+    Ret = iconv(ConvDesc, &Input, &InputLength, &Output, &OutputLength);
+    if (Ret != 0) {
+      if (auto EC = HandleError(Ret))
+        return EC;
+      continue;
+    }
+    // Flush the converter
+    Ret = iconv(ConvDesc, nullptr, nullptr, &Output, &OutputLength);
+    if (Ret != 0) {
+      if (auto EC = HandleError(Ret))
+        return EC;
+      continue;
+    }
+    break;
+  } while (true);
+
+  // Re-adjust size to actual size.
+  Result.resize(Output - Result.data());
+  return std::error_code();
+}
+
+void TextEncodingConverterIconv::reset() {
+  iconv(ConvDesc, nullptr, nullptr, nullptr, nullptr);
+}
+
+#endif // HAVE_ICONV
+} // namespace
+
+ErrorOr<TextEncodingConverter>
+TextEncodingConverter::create(TextEncoding CPFrom, TextEncoding CPTo) {
+
+  // Text encodings should be distinct.
+  if (CPFrom == CPTo)
+    return std::make_error_code(std::errc::invalid_argument);
+
+  ConversionType Conversion;
+  if (CPFrom == TextEncoding::UTF8 && CPTo == TextEncoding::IBM1047)
+    Conversion = UTF8ToIBM1047;
+  else if (CPFrom == TextEncoding::IBM1047 && CPTo == TextEncoding::UTF8)
+    Conversion = IBM1047ToUTF8;
+  else
+    return std::error_code(errno, std::generic_category());
+
+  return TextEncodingConverter(
+      std::make_unique<TextEncodingConverterTable>(Conversion));
+}
+
+ErrorOr<TextEncodingConverter> TextEncodingConverter::create(StringRef From,
+                                                             StringRef To) {
+  std::optional<TextEncoding> FromEncoding = getKnownEncoding(From);
+  std::optional<TextEncoding> ToEncoding = getKnownEncoding(To);
+  if (FromEncoding && ToEncoding) {
+    ErrorOr<TextEncodingConverter> Converter =
+        create(*FromEncoding, *ToEncoding);
+    if (Converter)
+      return Converter;
+  }
+#if HAVE_ICU
+  UErrorCode EC = U_ZERO_ERROR;
+  UConverterUniquePtr FromConvDesc(ucnv_open(From.str().c_str(), &EC));
+  if (U_FAILURE(EC)) {
```

Comment:
Style nit: Coding guidelines say no braces for the single-statement `if` block here.